hs/src/Reach/Simulator.hs

```diff
@@ -0,0 +1,477 @@
+      evd_args <- mapM interp dlargs
+      interpPrim (primop,evd_args)
+    DLE_ArrayRef _at dlarg1 dlarg2 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      case (ev1,ev2) of
+        (V_Array arr, V_UInt n) -> return $ arr !! (fromIntegral n)
+        _ -> impossible "expression interpreter"
+    DLE_ArraySet _at dlarg1 dlarg2 dlarg3 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      ev3 <- interp dlarg3
+      case (ev1,ev2) of
+        (V_Array arr, V_UInt n) -> do
+          let n' = fromIntegral n
+          return $ V_Array $ take n' arr ++ [ev3] ++ drop (n' + 1) arr
+        _ -> impossible "expression interpreter"
+    DLE_ArrayConcat _at dlarg1 dlarg2 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      case (ev1,ev2) of
+        (V_Array arr1, V_Array arr2) -> return $ V_Array $ arr1 ++ arr2
+        _ -> impossible "expression interpreter"
+    DLE_ArrayZip _at dlarg1 dlarg2 -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      case (ev1,ev2) of
+        (V_Array arr1, V_Array arr2) -> do
+          return $ V_Array $ map (\(l,r)-> V_Tuple $ [l,r]) $ zip arr1 arr2
+        _ -> impossible "expression interpreter"
+    DLE_TupleRef _at dlarg n -> do
+      ev <- interp dlarg
+      case ev of
+        V_Tuple arr -> return $ arr !! (fromIntegral n)
+        _ -> impossible "expression interpreter"
+    DLE_ObjectRef _at dlarg str -> do
+      ev <- interp dlarg
+      case ev of
+        V_Object obj -> return $ obj M.! str
+        _ -> impossible "expression interpreter"
+    DLE_Interact _at _slcxtframes _slpart _str _dltype dlargs -> do
+      args <- mapM interp dlargs
+      suspend $ PS_Suspend (Action_Interact args)
+    DLE_Digest _at dlargs -> V_Digest <$> V_Tuple <$> mapM interp dlargs
+    -- TODO
+    DLE_Claim _at _slcxtframes claimtype dlarg _maybe_bytestring -> case claimtype of
+      CT_Assert -> interp dlarg
+      CT_Assume _bool -> interp dlarg
+      CT_Require -> interp dlarg
+      CT_Possible -> return V_Null
+      CT_Unknowable _slpart _dlargs -> return $ V_Null
+    DLE_Transfer _at dlarg1 dlarg2 maybe_dlarg -> do
+      ev1 <- interp dlarg1
+      ev2 <- interp dlarg2
+      case (ev1,ev2) of
+        -- TODO: "Remove funds from contract..." which contract?
+        (V_UInt n, V_Address acc) -> do
+          case maybe_dlarg of
+            Nothing -> updateLedger acc 0 (+n) $ return V_Null
+            Just tok -> do
+              ev <- interp tok
+              case ev of
+                V_UInt tok' -> updateLedger acc tok' (+n) $ return V_Null
+                _ -> impossible "unexpected error"
+        _ -> impossible "expression interpreter"
+    DLE_TokenInit _at _dlarg -> return V_Null
+    DLE_CheckPay _at _slcxtframes dlarg _maybe_dlarg -> interp dlarg
+    DLE_Wait _at dltimearg -> case dltimearg of
+      Left dlarg -> do
+        ev <- interp dlarg
+        case ev of
+          V_UInt n -> incrNWtime n $ return V_Null
```

Comment:
This should actually be a blocking moment, because things might happen while someone or the contract is waiting for a certain time. Also `n` is not an AMOUNT to wait it is a TIME TO WAIT UNTIL